File: scala/top_model_mixin.scala
import scala.collection.mutable
import scala.util.Try

class LoraConfig
class Mapping
class PluginConfig {
  def fromDict(kwargs: mutable.Map[String, Any]): PluginConfig = {
    new PluginConfig()
  }
}

trait TopModelMixin {
  
  def fromHuggingFace(
    hfModelDir: String,
    dtype: Option[String] = Some("float16"),
    mapping: Option[Mapping] = None,
    kwargs: mutable.Map[String, Any] = mutable.Map()
  ): Unit = {
    throw new NotImplementedError("Subclass shall override this")
  }

  def useLora(loraConfig: LoraConfig): Unit = {
    throw new NotImplementedError("Subclass shall override this")
  }

  def usePromptTuning(
    maxPromptEmbeddingTableSize: String,
    promptTablePath: String
  ): Unit = {
    throw new NotImplementedError("Subclass shall override this")
  }

  def defaultPluginConfig(kwargs: mutable.Map[String, Any] = mutable.Map()): PluginConfig = {
    new PluginConfig().fromDict(kwargs)
  }
}

File: python/top_model_mixin.py
from typing import Optional

from .lora_manager import LoraConfig
from .mapping import Mapping
from .plugin.plugin import PluginConfig

class TopModelMixin:

    def __init__(self) -> None:
        pass

    @classmethod
    def from_hugging_face(cls,
                          hf_model_dir: str,
                          dtype: Optional[str] = 'float16',
                          mapping: Optional[Mapping] = None,
                          **kwargs):
        raise NotImplementedError("Subclass shall override this")

    def use_lora(self, lora_config: LoraConfig):
        raise NotImplementedError("Subclass shall override this")

    def use_prompt_tuning(self, max_prompt_embedding_table_size: str,
                          prompt_table_path: str):
        raise NotImplementedError

    def default_plugin_config(self, **kwargs) -> PluginConfig:
        return PluginConfig.from_dict(kwargs)
